File: crates/context/src/assemble.rs
//! §8.1 first stage + §8.5 budget degradation. Pure, deterministic, no I/O:
//! same state + phase + budget ⇒ the same document, bit for bit (I14).

use kernel::{PhaseId, SectionId};

use crate::degrade::{degrade, withhold_oversized, BINARY_SHARE_DIVISOR};
use crate::state::{SectionSource, State};
use crate::types::{Budget, CompactionReport, Document, Fidelity, Part, Section};

/// Rough token cost of a part list: bytes/4, floor 1 per part (Spike C).
/// Not a tokenizer; good enough to make a budget bind deterministically.
pub(crate) fn cost(parts: &[Part]) -> u32 {
    parts
        .iter()
        .map(|p| {
            let bytes = match p {
                Part::Text { text } => text.len(),
                Part::Image { data_base64, .. } | Part::Audio { data_base64, .. } => {
                    data_base64.len()
                }
                Part::File { data_base64, .. } => data_base64.len(),
                Part::Fragment { html, .. } => html.len(),
            };
            (bytes / 4).max(1) as u32
        })
        .sum()
}

/// The parts a section contributes at `fid`, derived from its FULL parts
/// (each level derives from the original, never from the previous level).
/// Summarized uses the provider's precomputed summary when present, else a
/// mechanical char-boundary-safe truncation (Spike C friction 3).
pub(crate) fn effective_parts(src: &SectionSource, fid: Fidelity) -> Vec<Part> {
    let s = &src.section;
    match fid {
        Fidelity::Full => s.parts.clone(),
        Fidelity::Summarized => match &src.summary {
            Some(parts) => parts.clone(),
            None => vec![Part::Text {
                text: mechanical_summary(s),
            }],
        },
        Fidelity::Pointer => vec![Part::Text {
            text: format!(
                "[section '{}': {} part(s) available; ask for them]",
                s.id.0,
                s.parts.len()
            ),
        }],
        Fidelity::Elided => Vec::new(),
    }
}

/// Leading text of the FULL parts, char-boundary safe, with a notice of what
/// was withheld — deterministic, no model in the loop (ADR-009).
fn mechanical_summary(s: &Section) -> String {
    const KEEP: usize = 200;
    let joined: String = s
        .parts
        .iter()
        .filter_map(|p| match p {
            Part::Text { text } => Some(text.as_str()),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join(" ");
    let head: String = joined.chars().take(KEEP).collect();
    let non_text = s
        .parts
        .iter()
        .filter(|p| !matches!(p, Part::Text { .. }))
        .count();
    format!(
        "{head} ...[summarized; {non_text} non-text part(s) withheld; ask for full '{}']",
        s.id.0
    )
}

/// Every section, at the fidelity it starts from, ordered for the prompt.
///
/// A section starts at Full with its real (recomputed) cost — except one with
/// nothing to say, which starts Elided. A component that does not apply this
/// turn (no stage brief, no observations yet) vanishes rather than rendering
/// an empty heading: Elided is how the paper already spells "absent", and it
/// is the level at which empty IS the content, so the law holds without an
/// exception written for it.
///
/// Order is structural: the slot decides, and nothing else. The sort is
/// stable, so two components sharing a slot keep the order they were supplied
/// in — deliberately NOT tie-broken on `Section::priority`, which is the
/// budget rank. Letting a budget number reorder the prompt would be the same
/// category error the slot type was introduced to end.
///
/// An oversized binary part is withheld here, before any ladder runs: charged
/// at bytes/4 a screenshot outweighs the conversation the ladder would
/// otherwise shred to make room for it.
fn gather(
    state: &State,
    ceiling: u32,
    withheld: &mut Vec<SectionId>,
) -> Vec<(SectionSource, Fidelity, u32)> {
    let mut work: Vec<(SectionSource, Fidelity, u32)> = state
        .sources
        .iter()
        .map(|src| {
            let mut src = src.clone();
            withhold_oversized(&mut src, ceiling, withheld);
            let c = cost(&src.section.parts); // >= 1 per part: 0 means empty
            let start = match c {
                0 => Fidelity::Elided,
                _ => Fidelity::Full,
            };
            (src, start, c)
        })
        .collect();
    work.sort_by_key(|(src, _, _)| src.section.slot);
    work
}

/// Build the paper for one call — the frozen §8.1 signature: gather every
/// section at its starting fidelity in slot order, let the budget take what it
/// must (`degrade`), and emit the document with a receipt for both. Total, not
/// fallible: malformed sections are rejected at module install time (ADR-004).
/// Sections carry their EFFECTIVE parts at the chosen fidelity: the document
/// is what the model sees, never a full-fidelity intermediate (ADR-009 Option
/// C's lie).
pub fn assemble(state: &State, phase: PhaseId, budget: Budget) -> Document {
    let mut withheld: Vec<SectionId> = Vec::new();
    let ceiling = budget.max_tokens / BINARY_SHARE_DIVISOR;
    let mut work = gather(state, ceiling, &mut withheld);
    let steps = degrade(&mut work, budget);
    let spent = work.iter().map(|(_, _, c)| *c).sum();
    let sections = work
        .into_iter()
        .map(|(src, fid, c)| {
            let mut s = src.section.clone();
            s.parts = effective_parts(&src, fid);
            s.fidelity = fid;
            s.budget_hint = c;
            s
        })
        .collect();
    Document {
        phase,
        sections,
        report: CompactionReport {
            budget,
            spent,
            steps,
            withheld,
        },
    }
}

File: packages/context/src/assemble.js
/**
 * The paper, assembled: sections in slot order, the pinned ends enforced, the
 * budget applied, and a receipt for everything it took (I13, I14).
 *
 * DETERMINISTIC. The same state and the same budget produce a byte-identical
 * document — no clock, no randomness, no model call, and the untrusted
 * envelope's nonce is derived from the payload for exactly that reason. A
 * golden test asserts it rather than the doc comment claiming it.
 *
 * There is no second constructor. `validate` is not exported from this
 * package and runs here, on the way out, so an invalid document cannot be
 * obtained at all.
 * @module
 */

import { estimateParts } from './estimate.js'
import { effectiveParts } from './fit.js'
import { nonceFor, wrapUntrusted } from './envelope.js'
import { nextFidelity, FIDELITIES } from './types.js'
import { validate, mentions } from './law.js'
import { withholdOversized, BINARY_SHARE } from './withhold.js'

/** @typedef {import('./types.js').Budget} Budget */
/** @typedef {import('./types.js').Document} Document */
/** @typedef {import('./types.js').Fidelity} Fidelity */
/** @typedef {import('./types.js').CompactionStep} CompactionStep */
/** @typedef {import('./state.js').State} State */
/** @typedef {import('./state.js').SectionSource} SectionSource */
/**
 * One section, mid-assembly. `parts` is carried rather than recomputed at the
 * end because the budget was spent against THESE bytes: recomputing them
 * later, when the allowance has moved, would let `budgetHint` describe a
 * document nobody assembled.
 * @typedef {{source: SectionSource, fidelity: Fidelity, tokens: number, parts: Part[]}} Fitted
 */
/** @typedef {import('./types.js').Part} Part */

/**
 * Build the paper for one call.
 * @param {State} state
 * @param {Budget} budget
 * @returns {Document}
 * @throws {HarnessError} by law name — `no_head`, `elided_but_named`, …
 */
export function assemble(state, budget) {
  /** @type {import('@harness/kernel').SectionId[]} */
  const withheld = []
  const work = gather(state, budget, withheld)
  const steps = degrade(work, budget)
  const doc = {
    stage: state.stage,
    sections: work.map(({ source, fidelity, tokens, parts }) => ({
      ...source.section,
      fidelity,
      budgetHint: tokens,
      parts,
    })),
    report: { budget, spent: total(work), steps, withheld },
  }
  validate(doc)
  return doc
}

/**
 * Every section at the fidelity it starts from, in prompt order.
 *
 * The sort is by slot and NOTHING else, and it is stable, so two sections
 * sharing a slot keep the order they were supplied in. Deliberately not
 * tie-broken on `priority`: that is the budget rank, and letting a budget
 * number reorder the prompt is the category error the slot type ended.
 * @param {State} state @param {Budget} budget
 * @param {import('@harness/kernel').SectionId[]} withheld
 * @returns {Fitted[]}
 */
function gather(state, budget, withheld) {
  const ceiling = Math.floor(budget.maxTokens / BINARY_SHARE)
  const nonce = nonceFor(state.sources.map((s) => s.section.id).join('|'))
  const ordered = [...state.sources].sort((a, b) => a.section.slot - b.section.slot)
  const referenced = referencedIn(ordered)
  return ordered.map((raw) => {
    const source = envelop(withholdOversized(raw, ceiling, withheld), nonce)
    const empty = source.section.parts.length === 0
    const named = referenced.has(source.section.id)
    const floor = startingFloor(source.section.floor, empty, named)
    const fidelity = empty ? (named ? 'pointer' : 'elided') : 'full'
    const parts = effectiveParts(source, fidelity, budget.maxTokens)
    const scoped = { ...source, section: { ...source.section, floor } }
    return { source: scoped, fidelity, parts, tokens: estimateParts(parts).tokens }
  })
}

/**
 * How far a section may be degraded, once the paper knows two things about it.
 *
 * A section another section's prose NAMES may never reach `elided`: that is
 * the `## observations` defect, where the operating rules told the model to
 * read a block the budget had removed. It stops at `pointer`, which says the
 * block exists and is being held back.
 *
 * An EMPTY section's floor is `elided` whatever it declared. A floor is a
 * statement about how much may be taken away, and there is nothing here to
 * take; the alternative is a document that fails its own law for having
 * nothing to say.
 * @param {Fidelity} declared @param {boolean} empty @param {boolean} named
 * @returns {Fidelity}
 */
function startingFloor(declared, empty, named) {
  if (empty) return named ? 'pointer' : 'elided'
  if (!named) return declared
  return FIDELITIES.indexOf(declared) > FIDELITIES.indexOf('pointer') ? 'pointer' : declared
}

/** Sections some other section's prose sends the model to read. @param {SectionSource[]} sources */
function referencedIn(sources) {
  /** @type {Set<string>} */
  const named = new Set()
  for (const target of sources) {
    if (sources.some((s) => s.section.id !== target.section.id && mentions(s.section, target.section.id))) {
      named.add(target.section.id)
    }
  }
  return named
}

/** Untrusted content, wrapped so nothing in it can be read as an instruction. @param {SectionSource} src @param {string} nonce */
function envelop(src, nonce) {
  if (src.section.trust !== 'untrusted') return src
  return {
    section: { ...src.section, parts: wrapUntrusted(src.section.parts, nonce) },
    summary: src.summary === null ? null : wrapUntrusted(src.summary, nonce),
  }
}

/**
 * While the paper is over budget, step the highest priority number not yet at
 * its floor down ONE level and recompute its cost against what is left for it
 * — the allowance is continuous, so a transcript drops the turns it has to
 * rather than collapsing to a fixed number of characters.
 *
 * Stops when everything sits at its floor. The honest overshoot is then in
 * the report rather than forced away by a cut nobody recorded.
 * @param {Fitted[]} work @param {Budget} budget
 * @returns {CompactionStep[]}
 */
function degrade(work, budget) {
  /** @type {CompactionStep[]} */
  const steps = []
  while (total(work) > budget.maxTokens) {
    const at = nextVictim(work)
    if (at === null) return steps
    const w = /** @type {Fitted} */ (work[at])
    const from = w.fidelity
    const to = nextFidelity(from)
    if (to === null) return steps
    w.fidelity = to
    w.parts = effectiveParts(w.source, to, allowanceFor(work, budget, w.tokens))
    w.tokens = estimateParts(w.parts).tokens
    steps.push({ section: w.source.section.id, from, to })
  }
  return steps
}

/**
 * What is left for one section once every other section has been paid for.
 * Never negative: at zero the primitives keep their minimum — the newest turn,
 * two characters — which is what makes an impossible budget produce a small
 * document rather than an empty one.
 * @param {Fitted[]} work @param {Budget} budget @param {number} mine
 */
function allowanceFor(work, budget, mine) {
  return Math.max(0, budget.maxTokens - (total(work) - mine))
}

/** The lowest-ranked section still above its floor; ties go to the later one. @param {Fitted[]} work */
function nextVictim(work) {
  let at = null
  for (let i = 0; i < work.length; i += 1) {
    const w = /** @type {Fitted} */ (work[i])
    if (FIDELITIES.indexOf(w.fidelity) >= FIDELITIES.indexOf(w.source.section.floor)) continue
    const best = at === null ? null : /** @type {Fitted} */ (work[at])
    if (best === null || w.source.section.priority >= best.source.section.priority) at = i
  }
  return at
}

/** @param {Fitted[]} work */
function total(work) {
  return work.reduce((sum, w) => sum + w.tokens, 0)
}
